src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1373,47 +1438,377 @@ public override bool Equals(object obj)
                 throw ExceptionUtilities.Unreachable;
             }
 
-            public bool Equals(RemainingTestsForCase other)
+            public bool Equals(StateForCase other)
             {
                 // We do not include Syntax, Bindings, WhereClause, or CaseLabel
                 // because once the Index is the same, those must be the same too.
-                return other != null && this.Index == other.Index && this.RemainingTests.SequenceEqual(other.RemainingTests, SameTest);
+                return other != null &&
+                    this.Index == other.Index &&
+                    this.RemainingTests.Equals(other.RemainingTests);
             }
 
-            private static bool SameTest(BoundDagTest x, BoundDagTest y)
+            public override int GetHashCode()
             {
-                if (x.Kind != y.Kind || x.Input != y.Input)
+                return Hash.Combine(RemainingTests.GetHashCode(), Index);
+            }
+        }
+
+        /// <summary>
+        /// A set of tests to be performed.  This is a discriminated union; see the options (nested types) for more details.
+        /// </summary>
+        private abstract class Tests
+        {
```

Comment:
The principal change in this source file is a change in representation of the tests remaining for a case.  It used to be an array of  `BoundDagTest`, with the implicit understanding that all tests in the array must succeed.  We change to a more explicit representation.  We have an `AndSequence` to represent the previous most common case.  `True` and `False` are used to represent a test that is unconditionally true or false without any code needed.  `OrSequence` is a set of tests any one of which succeeding means the whole thing succeeds.  `Not` is an inversion of the tests beneath it.  Finally, `One` represents a single test represented by the `BoundDagTest` that we had previously used as an element of the array.